users/views.py

```diff
@@ -4,9 +4,15 @@
 from django.http            import JsonResponse
 from django.db.utils        import IntegrityError
 from django.forms           import ValidationError
+from django.core            import mail
 from django.core.validators import validate_email
 from django.core.exceptions import ObjectDoesNotExist
 from django.contrib.auth    import login, authenticate
+from django.contrib.auth.tokens import default_token_generator
+from django.contrib.sites.shortcuts import get_current_site  
+from django.template.loader import render_to_string  
+from django.utils.encoding import force_bytes, force_str
+from django.utils.http import urlsafe_base64_encode, urlsafe_base64_decode
```

Comment:
did you end up finding out what these functions do?

Reply:
Yes! Encoding requires the the variable to be in byte form, and while after decoding, it needs to be reaffirmed into a string. I don't know the actual algorithm behind it, but it seems to implement python's encoding/decoding function with some other levels of further implementations with non-string characters. The source code is a little confusing to decipher.